File: src/main/scala/AssetManager.java
package com.github.jarlah.scalagraphics;

import javax.imageio.ImageIO;
import java.awt.*;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.Map;

public class AssetManager {
    private final Map<String, Image> assets = new HashMap<>();

    public Image getImage(String assetName) {
        if (!assets.containsKey(assetName)) {
            assets.put(assetName, loadImage(assetName));
        }
        return assets.get(assetName);
    }

    public Image getScaledImage(String assetName, Dimension dimension) {
        var image = getImage(assetName);
        var originalSize = new Dimension(image.getWidth(null), image.getHeight(null));
        var newSize = getScaledDimension(originalSize, dimension);
        return image.getScaledInstance(newSize.width, newSize.height, Image.SCALE_SMOOTH);
    }

    private Dimension getScaledDimension(Dimension imageSize, Dimension boundary) {
        double widthRatio = boundary.getWidth() / imageSize.getWidth();
        double heightRatio = boundary.getHeight() / imageSize.getHeight();
        double ratio = Math.max(widthRatio, heightRatio);

        return new Dimension((int) (imageSize.width * ratio), (int) (imageSize.height * ratio));
    }

    private Image loadImage(String assetName) {
        try {
            return ImageIO.read(getClass().getClassLoader().getResourceAsStream("assets/" + assetName));
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
    }
}

File: src/main/scala/AssetManager.scala
package com.github.jarlah.scalagraphics

import javax.imageio.ImageIO
import java.awt.*
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.util
import scala.collection.mutable
import scala.util.Try

class AssetManager {
  private val assets = mutable.Map[String, Image]()

  def getImage(assetName: String): Image = {
    if (!assets.contains(assetName)) {
      assets.put(assetName, loadImage(assetName))
    }
    assets(assetName)
  }

  def getScaledImage(assetName: String, dimension: Dimension): Image = {
    val image = getImage(assetName)
    val originalSize =
      new Dimension(image.getWidth(null), image.getHeight(null))
    val newSize = getScaledDimension(originalSize, dimension)
    image.getScaledInstance(newSize.width, newSize.height, Image.SCALE_SMOOTH)
  }

  private def getScaledDimension(imageSize: Dimension, boundary: Dimension) = {
    val widthRatio = boundary.getWidth / imageSize.getWidth
    val heightRatio = boundary.getHeight / imageSize.getHeight
    val ratio = Math.max(widthRatio, heightRatio)
    new Dimension(
      (imageSize.width * ratio).toInt,
      (imageSize.height * ratio).toInt
    )
  }

  private def loadImage(assetName: String) = Try(
    ImageIO.read(
      getClass.getClassLoader.getResourceAsStream("assets/" + assetName)
    )
  ).getOrElse(null)
}
